server/routes/oauth/index.ts

```diff
@@ -0,0 +1,122 @@
+import OAuth2Server from "@node-oauth/oauth2-server";
+import Koa from "koa";
+import bodyParser from "koa-body";
+import Router from "koa-router";
+import { ValidationError } from "@server/errors";
+import auth from "@server/middlewares/authentication";
+import { rateLimiter } from "@server/middlewares/rateLimiter";
+import requestTracer from "@server/middlewares/requestTracer";
+import { transaction } from "@server/middlewares/transaction";
+import validate from "@server/middlewares/validate";
+import { OAuthClient } from "@server/models";
+import OAuthAuthentication from "@server/models/oauth/OAuthAuthentication";
+import { authorize } from "@server/policies";
+import { APIContext } from "@server/types";
+import { RateLimiterStrategy } from "@server/utils/RateLimiter";
+import { OAuthInterface } from "@server/utils/oauth/OAuthInterface";
+import oauthErrorHandler from "./middlewares/oauthErrorHandler";
+import * as T from "./schema";
+
+const app = new Koa();
+const router = new Router();
+const oauth = new OAuth2Server({
+  model: OAuthInterface,
+});
+
+router.post(
+  "/authorize",
+  rateLimiter(RateLimiterStrategy.FiftyPerHour),
+  auth(),
+  async (ctx) => {
+    const { user } = ctx.state.auth;
+    const clientId = ctx.request.body.client_id;
+    if (!clientId) {
+      throw ValidationError("Missing client_id");
+    }
+
+    const client = await OAuthClient.findByClientId(clientId);
+    authorize(user, "read", client);
+
+    // Note: These objects are mutated by the OAuth2Server library
+    const request = new OAuth2Server.Request(ctx.request);
+    const response = new OAuth2Server.Response(ctx.response);
+
+    const authorizationCode = await oauth.authorize(request, response, {
+      allowEmptyState: true,
```

Comment:
What's the rationale for allowing empty state here? it seems like this reduces csrf protection?

Reply:
Generally trying to keep the implementation as flexible as possible – even Google does not _require_ the state parameter, it is left as a recommended option for the client to implement